### ISSUE TITLE
Compiling for size

### ISSUE THREAD

**Issue by stellaraccident** [google/iree#898]
We ultimately need to be compiling the runtime portions of IREE for size. To that end, I started fiddling with the CMake build to play with optimization options. It was hard/hacky to get it working at all, and doing it for real will likely require more build surgery.

These experiments were all done on iree-run-module on x86_64 linux. This is not a perfect proxy because:
  - It contains significant IO and string manipulation that would not be present in a real API user
  - It is a standalone binary vs library
  - It has a lot of debug and safety facilities compiled in.

Still, this largely just means that it is not predictive as to the floor of how small we can make things, but it does provide some relative benefit when comparing options and looking at sorted list of size offenders.

### Important Flags:

* `-flto=thin`: Enables link time, whole program optimizations. Really hard to get right, especially with respect to whole-archive related things. Anecdotally, this can decrease the size of our binary by ~20-30%.
* `-fnoexceptions`: Disables language features that require bulky eh_frame sections
* `-fno-unwind-tables -fno-asynchronous-unwind-tables`: Disables frame unwind info (assuming no language features use it). Saves up to 50% of the binary size if applied to everything (at the expense of debugability).

### General observations:

* Abseil does not respect global C flags and uses its own. I ended up hacking their CMake files to do exactly what I wanted (the above). Specifically, they don't build with any of the above by default. Abseil alone was contributing ~314KiB of unwind sections before I got all of them.
* LTO mode is really tricky to use and is largely incompatible with static archives use in whole archive regions. I expect that for when it matters, we are going to want dedicated targets that produce LTO binaries and only depend on LTO inputs. Especially with the LLVM and Abseil deps, this is really hard to get working globally.

### Current sizes:

I measured hacked builds of the following configurations:

* No HAL modules (`base`): 699KiB stripped binary size (>1.5MiB prior to build optimizations)
* Only the Vulkan HAL module (`vulkan`): 972KiB stripped binary size
* Only VMLA module (`vmla`): 987KiB stripped binary size
* Vulkan+VMLA modules (`vulkan_vmla`): 1.2MiB stripped binary size

I used Josh's [bloaty](https://github.com/google/bloaty) tool to get some size profiles (attached for each) with commands like: `bloaty iree-run-module -n 0 -d sections,symbols -s vm`

[base_bloaty.txt](https://github.com/google/iree/files/4271320/base_bloaty.txt)
[vulkan_bloaty.txt](https://github.com/google/iree/files/4271321/vulkan_bloaty.txt)
[vmla_bloaty.txt](https://github.com/google/iree/files/4271322/vmla_bloaty.txt)
[vulkan_vmla_bloaty.txt](https://github.com/google/iree/files/4271324/vulkan_vmla_bloaty.txt)

I had to hack on the build so much to even get these that it is not reproducible. Thanks to Marius for getting the CMake build to this point -- which made this possible!

**Comment by stellaraccident** [comment]
Note also that I couldn't get the LLVMJit compiling with these optimizations. Without them, it was weighing in at ~13MiB.

**Comment by GMNGeoffrey** [comment]
If all we care about is the runtime, I think we should be looking at iree-run-module instead of iree-run-mlir

**Comment by stellaraccident** [comment]
Yes, that's what I was looking at.

**Comment by GMNGeoffrey** [comment]
> Yes, that's what I was looking at (updated the typo in the issue that was referring to the wrong one). It is virtually impossible to confuse the two: iree-run-mlir is 10x larger.

Ok good :-D

**Comment by benvanik** [comment]
Yeah, we'll need to get a build of an executable that does not use absl flags/the buffer string parsing/etc. I'm also leaning more and more towards our own status that lets us easily strip things out; so much of the long tail in these is random status formatting junk.

I'm surprised that the packing dispatches aren't getting deduped. That's a shockingly large amount of code. I'll have to look into that.

**Comment by stellaraccident** [comment]
> I'm surprised that the packing dispatches aren't getting deduped. That's a shockingly large amount of code. I'll have to look into that.

That was the easter egg I left in there so I knew you opened the files :)

**Comment by stellaraccident** [comment]
> Yeah, we'll need to get a build of an executable that does not use absl flags/the buffer string parsing/etc. I'm also leaning more and more towards our own status that lets us easily strip things out; so much of the long tail in these is random status formatting junk.

Agreed. But the build surgery needed is a bit gross -- and can be done on iree-run-module as easily on a new binary.

**Comment by benvanik** [comment]
wait, but if iree-run-module is basically just flags and buffer parsing, wouldn't we need a new binary to not have flags or buffer parsing? :)

**Comment by stellaraccident** [comment]
> wait, but if iree-run-module is basically just flags and buffer parsing, wouldn't we need a new binary to not have flags or buffer parsing? :)

This might be a good time to create that C-simple API we've discussed, with a couple of easy entry points and package it up as a shared library. It'd be a good size test and something we need anyway.

**Comment by benvanik** [comment]
Since I only understand things when boxes are involved, here's iree-run-module on Windows (release build with optimize-for-speed):
![image](https://user-images.githubusercontent.com/75337/75639391-63e7de80-5be5-11ea-8fad-a2474322680b.png)

Of note:
- 600KB of STL/etc
- 200+KB for absl str_format & co (!)
- 60KB for absl::Mutex
- 100KB for InlinedVector inlined calls (likely inlined usage for all callsites)
- 130KB for statusor (also likely inlined)
- 70KB for buffer_string_util (the input/output buffer parsing)
- 100KB for flags
- 180KB for VMA (may be ways to get this smaller, there are a lot of flags)
- 35KB for Ruy (in whatever config tit's compiled in)
- VMLA+host thread queue with all kernels is 250KB, Vulkan+VMA is 380KB

Unfortunately the STL stuff is busted in the tool (probably space in file path), but looking at the symbolsort dump it looks like:
- 205KB for std::vector
- 80KB for memory
- 60KB for functional (std::function/bind/etc)
- probably some big annoying timezone table crap

Zooms of absl:
![image](https://user-images.githubusercontent.com/75337/75639524-11f38880-5be6-11ea-9121-84c3fb4a3c68.png)

Zoom of IREE:
![image](https://user-images.githubusercontent.com/75337/75639535-23d52b80-5be6-11ea-92f9-00b6fe938637.png)

I think there's likely some linker settings that could help here as 7zip takes the exe from 890KB to 300KB - so there's roughly ~300KB of useful unique data in there. I also haven't tried linking the CRT in shared library mode.

So, thinking:
- 12KB for the core VM interpreter is pretty nice
- ~200-300KB per HAL backend seems reasonable
- 800+KB for absl::InlinedVector/Mutex/string formatting seems excessive; may be time to excise more of that in favor of types that generate smaller code
- 200KB for all the status related machinery seems excessive, and more reason to go with the one I'm working on
- the 150KB of module_abi_packing is likely all the strings and error handling I just added - making that configurable would be nice (debug only/etc)

**Comment by benvanik** [comment]
symbolsort dump:
[iree-run-module.txt](https://github.com/google/iree/files/4273093/iree-run-module.txt)

**Comment by stellaraccident** [comment]
> * 12KB for the core VM interpreter is pretty nice
> * ~200-300KB per HAL backend seems reasonable
> * 800+KB for absl::InlinedVector/Mutex/string formatting seems excessive; may be time to excise more of that in favor of types that generate smaller code

We need to see this in a more minimal API-only build, I think, in order to make judgments. Dumping absl::Mutex, though, probably should be done regardless -- it's main value proposition is with respect to interop with some of Google's infra, which doesn't apply here. I've always maintained that for distribution to various devices, you want to use the synchronization mechanisms that the platform provides unless if it matters for specific code, in which case you want to be *very* opinionated about your choice and not just rely on another black box implementation not optimized towards what you want (which is what absl::Mutex is).

I believe that most of the bulky string/timezone stuff is coming in from either status, logging or the main() parts of iree-run-module which wouldn't go in an API build.

Also, keep in mind that unless if you did something, I think that ABSL is compiling with exceptions enabled and full unwind tables. In my early trials, this seemed to be causing ABSL to appear much larger than it actually was (relative to the other artifacts). Probably also defeats some deduping inlining and such due to ABI mismatches. Once I got that fixed and LTO enabled, the picture seemed more reasonable. I think the next step is to duplicate these measurements with a more minimal binary and making sure to have the flags set right on all targets.

> * 200KB for all the status related machinery seems excessive, and more reason to go with the one I'm working on

+1 - we knew this was coming :)

> * the 150KB of module_abi_packing is likely all the strings and error handling I just added - making that configurable would be nice (debug only/etc)

Not worried about it for debug features that we can/should be disabling with compile options. It would be good to verify.

I also noticed on the Linux build that function/data sections and gc are not enabled. We should make a pass through and tune the flags -- they basically have had no diligence put on them yet.

From my perspective, the next steps are something like this (I wouldn't put any of this at the top of the list, at the moment, as I'd rather have the core working well vs playing flag-golf for last mile build tuning):

1. Create a simple C-API library that knows how to a) initialize a context, b) load modules, c) invoke functions. I think there should be a really simple API for this (regardless of access to the lower-level, higher strength APIs).
1. Make sure that we have a place where we can set compile/link flags for all runtime parts of the CMake build (I may not have been doing it right, but I spent significant time trying to figure out why common-sense settings like CXXFLAGS or the requisite CMake var equivalents were not making their way to some targets).
1. Make sure that we can specify flags impacting the runtime components independent from the compiler and tools. I'm not sure how LLVM does it, but the way they treat tablegen as "special" so that it can be built optimized, even in an overall debug build is a variant of what I think we need here.
1. Make sure we have CMake options for conditionally compiling out verbose error handling, logging and status code (this will help us better estimate the "floor" for a lot of the string and other utilities that are just along for the ride in a verbose build).
1. Do some of the hygienic replacements you mention above. We don't need a lot of data to know that some of those things are not worth the bloat value.

**Comment by benvanik** [comment]
Yeah, I'll get the status/error stuff (#265) done and that'll get us a better baseline. I can also work on the minimal C API this week. Exceptions are on here but MSVC does a better job with them (hence my 800KB vs your 1.2MB), but still would be good to disable.

iree_copts.cmake should be where everything is set - if things aren't respecting them we should fix that. CMake does have a loose concept of host vs target that we can use as well.

I want us to be able to use user-provided types in most cases (particularly mutex and logging) and part of the work of factoring that out will let us either use abseil types inside of google and std/pthread/win32 elsewhere (and then let us more easily be adapted). Similar to what VMA does for external mutex support.

**Comment by antiagainst** [comment]
Sidetracking: Ben, how did you generate the nice boxes in https://github.com/google/iree/issues/898#issuecomment-593187460?

**Comment by benvanik** [comment]
symbolsort + bloat: https://github.com/rongjiecomputer/bloat-win
it's got some bugs, but was enough for some quick pictures :)

**Comment by benvanik** [comment]
I haven't tried it yet, but this fork had some stuff for detecting template-based bloat that I'd love to try: https://dirtyhandscoding.github.io/posts/improvements-to-symbolsort-tool-for-c-code-bloat-analysis.html

**Comment by benvanik** [comment]
Testing with an iree/vm-only executable that just loads and runs a dummy module, compiled with MSVC `/GL` (whole program optimization + link-time code gen):
[dump.zip](https://github.com/google/iree/files/4939335/dump.zip)

![image](https://user-images.githubusercontent.com/75337/87816258-e6cba780-c81b-11ea-8db1-379ee888eaa8.png)

Of note:
- abseil is 300KB for code that is almost purely C
- STL is 100KB mostly comprised of vector, xtree+hash/etc, functional, and string
- the iree/vm is 28KB, and almost all use of iree/base is logging/status/initialization (32KB)

absl/ zoom:
![image](https://user-images.githubusercontent.com/75337/87816274-ee8b4c00-c81b-11ea-834e-3411a09e2d97.png)

I'm deeply skeptical of the value derived from using absl::Duration/Time vs. std::chrono vs. simple uint64 values, especially because our use is entirely internal (we can't pass absl types across the C API). Also skeptical of the value of mutex, random junk pulled in from string code like int128, etc.

iree/ zoom:
![image](https://user-images.githubusercontent.com/75337/87816306-f64af080-c81b-11ea-9d47-107c8a26187b.png)

Even moreso this makes me want to cut the last remaining piece of C++ in iree/vm/, which is the bytecode module. From that point the entire VM should be < 30KB and our external deps are effectively just memcmp and memcpy.

**Comment by benvanik** [comment]
Latest snapshot of the VM size benchmark (x64, no HAL) while working on #265: 
![image](https://user-images.githubusercontent.com/75337/89561154-f0b64a00-d7cc-11ea-8ec8-af5640cf28e3.png)

Of note:
* the entire IREE VM to run arbitrary bytecode modules in flatbuffers is **25KB** and relies only on a few stdlib functions (strcmp/etc)
* the other half (~20KB) of the iree/ tree is in status/logging that will be removed
* abseil pulls in the world still (~250KB for itself, ~70KB for STL), but none of that is needed and is on the kill list
* flatcc builder.c is getting pulled in for 12KB - no clue why - will try to fix that with build configuration

Zoom of iree/:
![image](https://user-images.githubusercontent.com/75337/89561504-720ddc80-d7cd-11ea-97fc-c651ad51fa6c.png)

So, getting down to 25KB + HAL is pretty exciting. That puts us squarely in 64KB MCU territory and means running on accelerators even in bytecode form is possible.

I've been queuing up some work to tblgen shims for modules instead of using iree/vm/module_abi_cc.h for the HAL, which would likely add 10-20KB for all the verification/argument marshaling/etc when done. If running on an MCU/GPU/TPU the HAL C API could be implemented from scratch instead of using the larger/more flexible C++ HAL APIs, so I'm not concerned about that size too much. I'll still go through and de-abseil it all (use std::mutex instead of absl::Mutex, etc) so that we are not wasting space, but pulling in some STL is fine (as likely any application linking IREE into it already has the STL so the cost is amortized).

**Comment by benvanik** [comment]
(this was compiling with `-DNDEBUG` and `-DIREE_STATUS_MODE=0` which disables all status annotations)

**Comment by benvanik** [comment]
Continuing the glide, the size benchmark is now starting to settle in a good place:
![image](https://user-images.githubusercontent.com/75337/89704842-7db3ed00-d90c-11ea-9b5f-e22454a8b16e.png)

That's 27KB for the entire IREE bytecode VM. There's not getting much smaller than that and still doing what it is doing! :)

I'll track down the weird flatcc builder linkage tomorrow and Scott's #2823 should cut out the rest of abseil/the STL. At that point our total binary is <30KB in a size-optimized build.

Progress on the #265 feature branch is dramatically simplifying Status and cutting larger abseil deps (strings, flags, etc). Once that's in a better shape I'll look at getting a basic HAL benchmark (probably with the dylib backend as that measures just the HAL interface/module overhead and not the HAL implementation).

**Comment by benvanik** [comment]
Easier than I thought - proper linker configuration for function level linking and patching in scott's abseil time replacement nets a ~30KB x64 exe file that gzips to 17KB (that also includes the embedded a+b bytecode module flatbuffer): 
![image](https://user-images.githubusercontent.com/75337/89705205-37608d00-d910-11ea-9fd1-5f425e3188cd.png)

Less important than the sum size of the VM is what this means when the HAL/VMLA/etc is added: almost all of the binary size will be dominated by the HAL module glue and the actual implementation which is precisely the goal of IREE -- and that's with the full flatbuffer verifier and (securable but not yet fully secure) bytecode VM.

COPTS:
```
    "/GS-"
    "/GL"
    "/Gw"
    "/Gy"
    "/DNDEBUG"
    "/DIREE_STATUS_MODE=0"
```

LINKOPTS: `/LTCG /opt:ref,icf`

**Comment by benvanik** [comment]
hah, correction: disabling the i64 support in the VM drops us closer to 20KB:
![image](https://user-images.githubusercontent.com/75337/89705406-f49fb480-d911-11ea-890a-1199fcb23fb2.png)

**Comment by benvanik** [comment]
Quick glance at iree-run-module with HAL + dylib before bed:
![image](https://user-images.githubusercontent.com/75337/89705895-d1770400-d915-11ea-874f-b9b6f0e45e30.png)

Of note:
- logging and flags are not stripped - lots of strings in the binary from code doing validation
- module_abi_packing.h is a beast, as suspected
- hal api.cc and buffer.cc should be able to shrink
- HAL with async serial queue + dylib driver + validation is ~70KB and I suspect could easily go <50KB
- hal_module exposing HAL to the VM is 26KB + module_abi_packing.h - there's a lot of strings in there clouding things but I estimate it could hit 30-40KB total based on what it's doing
- rest of stuff is on the kill list or not required (only used in iree-run-module) like file IO/abseil/flags/etc

So right now we are about ~200KB of legit code + strings, and I suspect 100-150KB is within reach with no functional changes. The executable compresses at a 5x ratio which highlights there's a lot of redundancy (strings) or data that could be moved to bss.

Based on that, as expected for iree-run-module with HAL + VMLA, we gain decent chunk from the module_abi_packing.h used in vmla_module and the logging/strings in vmla_module:
![image](https://user-images.githubusercontent.com/75337/89706029-15b6d400-d917-11ea-8713-786e07d408ca.png)

- actual kernels for all types (int/float/etc) are ~76KB
- ruy is 16KB

Like above most of the bloat is shared and things that can be fixed (namely the module_abi_packing and vmla_module strings as with hal_module). I suspect we can hit ~100KB total inclusive of deps for VMLA.

So 200-250KB of x64 code as an upper bound for multithreaded IREE VM + HAL + VMLA that can load, verify, and execute arbitrary models is not bad and gives us plenty of room to add new features. 100-150KB for the same but with models using codegen (where code size is shifted to models) is also not bad. The only way to get smaller would be to codegen the VM, but at 20KB total and a higher code density than ARM/x64 it'd be exceedingly hard to beat as a total package size and is likely within ~small % of optimal.

🕺

**Comment by benvanik** [comment]
(also for size-constrained builds we can cut down VMLA ops by a significant margin - no need for i64/f32/etc)

**Comment by benvanik** [comment]
As part of this work we should set up a CI build for a minimum-size build (IREE_STATUS_MODE=0, IREE_LOGGING_MODE=0, etc).

**Comment by GMNGeoffrey** [comment]
> As part of this work we should set up a CI build for a minimum-size build (IREE_STATUS_MODE=0, IREE_LOGGING_MODE=0, etc).

I have added that to my TODO list (which is currently a literal piece of paper on my desk :-D). I intend to write down all these sorts of tasks and put them somewhere they can be easily picked off, but that probably won't happen for a couple weeks realistically.

**Comment by benvanik** [comment]
Update from a local branch where I'm cleaning up calling conventions/FFI. This reduces the module_abi_packing.h overhead by about half and shrinks small things in many other areas too (a few hundred of the thousand papercuts).

The size benchmark (full bytecode VM + flatbuffer module + flatbuffer verification) shows 25KB total + 3KB of CRT:
![image](https://user-images.githubusercontent.com/75337/91623977-b31f8980-e952-11ea-9e89-23cf18be2b84.png)

iree-run-module (VM, HAL, dylib backend, and string buffer parsing/formatting utils) is looking good too:
![image](https://user-images.githubusercontent.com/75337/91624124-5bcde900-e953-11ea-8601-52815da6e5fb.png)

This is still pulling in absl flags (which then pulls in synchronization and time utils) totaling 100KB + STL bits and using some non-prod HAL features and file IO (roughly 30-35KB). I'll see about creating a size benchmark without those things. This puts us at around 160-180KB for a full IREE CPU runtime. I'll continue iterating to reduce module_abi_packing.h, though an alternative (or additional) solution is to port the hal_module to C (which is good for portability and it's just a C -> C++ -> C so there's no need for it to be C++).

**Comment by benvanik** [comment]
No quick fixes for module_abi_packing unfortunately - I tried to do some `extern template` magic, however LTO takes care of whatever I was able to do with that anyway. That does mean that if more than a single module is using it we'll get the cost amortized across all of them for a lot of cases though it does set the min size for the entire binary. The only way to make that smaller is to switch to C and rely on cross-procedure optimizations (maybe I'll try that next friday, as it'll cause me to JIT design some nice C ABI utilities).
